File: src/main/scala/dev/alexnader/framity/client/assets/overlay/Offsetters.scala
package dev.alexnader.framity.client.assets.overlay

import dev.alexnader.framity.util.json.{JsonParseContext, JsonParseResult, JsonParser}

object Offsetters {

  implicit object Parser extends JsonParser[Offsetters] {
    override def apply(context: JsonParseContext): JsonParseResult[Offsetters] = {
      val uContext = context("uOffsetter")
      val vContext = context("vOffsetter")

      (uContext, vContext) match {
        case (Right(uContext), Right(vContext)) => uContext.parse[Offsetter].flatMap(uOffsetter => vContext.parse[Offsetter].map(vOffsetter => UV(uOffsetter, vOffsetter)))
        case (Right(uContext), Left(_)) => uContext.parse[Offsetter].map(U.apply)
        case (Left(_), Right(vContext)) => vContext.parse[Offsetter].map(V.apply)
        case (Left(_), Left(_)) => Left(context.makeFailure("Expected key(s) 'uOffsetter', 'vOffsetter', or both."))
      }
    }
  }

  case class U(uOffsetter: Offsetter) extends Offsetters {
    override def apply(us: (Float, Float, Float, Float), vs: (Float, Float, Float, Float)): ((Float, Float, Float, Float), (Float, Float, Float, Float)) = (uOffsetter(us), vs)
  }

  case class V(vOffsetter: Offsetter) extends Offsetters {
    override def apply(us: (Float, Float, Float, Float), vs: (Float, Float, Float, Float)): ((Float, Float, Float, Float), (Float, Float, Float, Float)) = (us, vOffsetter(vs))
  }

  case class UV(uOffsetter: Offsetter, vOffsetter: Offsetter) extends Offsetters {
    override def apply(us: (Float, Float, Float, Float), vs: (Float, Float, Float, Float)): ((Float, Float, Float, Float), (Float, Float, Float, Float)) = (uOffsetter(us), vOffsetter(vs))
  }

}

sealed abstract class Offsetters {
  def apply(us: (Float, Float, Float, Float), vs: (Float, Float, Float, Float)): ((Float, Float, Float, Float), (Float, Float, Float, Float))
}

File: src/main/java/dev/alexnader/framity/client/assets/overlay/Offsetters.java
package dev.alexnader.framity.client.assets.overlay;

import com.mojang.serialization.Codec;
import com.mojang.serialization.codecs.RecordCodecBuilder;
import net.fabricmc.api.EnvType;
import net.fabricmc.api.Environment;

import javax.annotation.Nonnull;
import java.util.Optional;

import static dev.alexnader.framity.client.FramityClient.CODECS;

@Environment(EnvType.CLIENT)
public class Offsetters {
    public static final Codec<Offsetters> CODEC = RecordCodecBuilder.create(inst -> inst.group(
        CODECS.OFFSETTER.optionalFieldOf("u").forGetter(o -> o.u.toOptional()),
        CODECS.OFFSETTER.optionalFieldOf("v").forGetter(o -> o.v.toOptional())
    ).apply(inst, Offsetters::new));

    public static final Offsetters NONE = new Offsetters();

    public final @Nonnull Offsetter u;
    public final @Nonnull Offsetter v;

    @SuppressWarnings("OptionalUsedAsFieldOrParameterType")
    public Offsetters(@Nonnull final Optional<Offsetter> u, @Nonnull final Optional<Offsetter> v) {
        this.u = u.orElse(Offsetter.NONE);
        this.v = v.orElse(Offsetter.NONE);
    }

    private Offsetters() {
        this.u = Offsetter.NONE;
        this.v = Offsetter.NONE;
    }
}
